File: src/lib/commands/task.ts
import type { TaskInfo, TaskType } from '../../types/task';
import { apiGet, apiPost } from '../api/client';

export async function createTask(
  worktreePath: string,
  taskType: TaskType,
  repo: string,
  branch: string,
  pr?: number,
  description?: string,
): Promise<TaskInfo> {
  try {
    return await apiPost<TaskInfo>('/tasks', {
      worktreePath,
      taskType,
      repo,
      branch,
      pr: pr ?? null,
      description: description || null,
    });
  } catch (e) {
    console.error(`createTask failed: ${e}`);
    throw e;
  }
}

export async function getTask(worktreePath: string): Promise<TaskInfo> {
  try {
    return await apiGet<TaskInfo>(`/tasks/detail?worktreePath=${encodeURIComponent(worktreePath)}`);
  } catch (e) {
    console.error(`getTask failed: ${e}`);
    throw e;
  }
}

export async function listTasks(worktreePaths: string[]): Promise<TaskInfo[]> {
  try {
    return await apiPost<TaskInfo[]>('/tasks/list', { worktreePaths });
  } catch (e) {
    console.error(`listTasks failed: ${e}`);
    throw e;
  }
}

export async function startTaskWatcher(worktreePaths: string[]): Promise<void> {
  try {
    await apiPost('/tasks/watcher/start', { worktreePaths });
  } catch (e) {
    console.error(`startTaskWatcher failed: ${e}`);
    throw e;
  }
}

export async function stopTaskWatcher(): Promise<void> {
  try {
    await apiPost('/tasks/watcher/stop');
  } catch (e) {
    console.error(`stopTaskWatcher failed: ${e}`);
    throw e;
  }
}

export async function watchTaskPath(worktreePath: string): Promise<boolean> {
  try {
    return await apiPost<boolean>('/tasks/watcher/add', { worktreePath });
  } catch (e) {
    console.error(`watchTaskPath failed: ${e}`);
    throw e;
  }
}

File: src-tauri/src/commands/task.rs
use crate::error::AppError;
use crate::models::task::{TaskInfo, TaskType};
use crate::services::task_watcher::TaskWatcherState;
use tauri::State;

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn create_task_cmd(
    worktree_path: String,
    task_type: TaskType,
    repo: String,
    branch: String,
    pr: Option<u64>,
    description: Option<String>,
) -> Result<TaskInfo, AppError> {
    crate::services::task::create_task(
        &worktree_path,
        task_type,
        &repo,
        &branch,
        pr,
        description.as_deref(),
    )
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn get_task_cmd(worktree_path: String) -> Result<TaskInfo, AppError> {
    crate::services::task::get_task(&worktree_path)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn list_tasks_cmd(worktree_paths: Vec<String>) -> Result<Vec<TaskInfo>, AppError> {
    crate::services::task::list_tasks(&worktree_paths)
}

#[tauri::command]
pub async fn start_task_watcher(
    watcher: State<'_, TaskWatcherState>,
    app_handle: tauri::AppHandle,
    worktree_paths: Vec<String>,
) -> Result<(), AppError> {
    let mut w = watcher.lock().await;
    w.start(&app_handle, &worktree_paths)
}

#[tauri::command]
pub async fn stop_task_watcher(watcher: State<'_, TaskWatcherState>) -> Result<(), AppError> {
    let mut w = watcher.lock().await;
    w.stop();
    Ok(())
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub async fn watch_task_path(
    watcher: State<'_, TaskWatcherState>,
    worktree_path: String,
) -> Result<bool, AppError> {
    let mut w = watcher.lock().await;
    w.watch_path(&worktree_path)
}
